packages/mfsu/src/utils/webpackUtils.ts

```diff
@@ -27,3 +30,63 @@ export function extractBabelPluginImportOptions(
 
   return configs;
 }
+
+export function tryEntryResolve(
+  entry: string,
+  basedir: string = process.cwd(),
+): string | undefined {
+  if (isRelativePath(entry)) {
+    const fsPath = path.resolve(basedir, entry);
+    return tryResolve(fsPath);
+  }
+
+  if (path.isAbsolute(entry)) {
+    return tryResolve(entry);
+  }
+}
+
+function tryResolve(absoluteEntry: string): string | undefined {
+  let res: string | undefined;
+  const extensions = ['.tsx', '.ts', '.jsx', '.js'];
+
+  // path/to/dir/main.[ext]
+  if ((res = tryResolveFile(absoluteEntry))) {
+    return res;
+  }
+
+  // path/to/dir/main -> path/to/dir/main.[ext]
+  for (let ext of extensions) {
+    if ((res = tryResolveFile(absoluteEntry + ext))) {
+      return res;
+    }
+  }
+
+  // path/to/dir/ -> path/to/dir/index.[ext]
+  for (let ext of extensions) {
+    if ((res = tryResolveFile(path.join(absoluteEntry, 'index' + ext)))) {
+      return res;
+    }
+  }
+}
+
+function tryResolveFile(absoluteEntry: string) {
+  if (
+    isFileReadable(absoluteEntry) &&
+    !fs.statSync(absoluteEntry).isDirectory()
+  ) {
+    return absoluteEntry;
+  }
+}
+
+function isRelativePath(fsPath: string) {
+  return RELATIVE_PATH_REGEXP.test(fsPath);
+}
+
+export function isFileReadable(filename: string): boolean {
```

Comment:
没用到的函数没必要 `export` ，大部分函数的返回类型可以自动推断，没必要特别声明（上面大部分函数的返回类型都可以不写。）

Reply:
写函数返回值的声明是因为之前在《Effective Typescript 》看到这样的解释：

> Consider using explicit annotations for object literals and function return types even when they can be inferred. This will help prevent implementation errors from surfacing in user code - 《Effective Typescript》Item 19

并且看到 Vite 源码也喜欢写明返回值，比如一个不返回的也会指名为 `void`：

```js
export function copyDir(srcDir: string, destDir: string): void {
  ...
}
```